source/agora/node/Ledger.d

```diff
@@ -494,6 +506,56 @@ public class Ledger
         return null;
     }
 
+    /***************************************************************************
+
+        Check whether the slashing data is valid.
+
+        Params:
+            data = consensus data
+
+        Returns:
+            the error message if validation failed, otherwise null
+
+    ***************************************************************************/
+
+    public string validateSlashingData (ConsensusData data) @trusted
```

Comment:
`in ConsensusData data`

Reply:
Can we make this function `@safe` and also mark `isInvalidPreimageRootReason` as `@safe` too if needed? We should avoid labeling everything as `@trusted`, because it tells the compiler to avoid doing any kind of safety checks.

Reply:
> Can we make this function @safe and also mark isInvalidPreimageRootReason as @safe too if needed?

We can. I'm not gonna use @trust as far as possible.